critwall/src/main/java/org/opencadc/critwall/FileSyncJob.java

```diff
@@ -0,0 +1,212 @@
+
+    private final ArtifactDAO artifactDAO;
+    private final URI artifactID;
+    private final URI resourceID;
+    private final StorageAdapter storageAdapter;
+
+    public FileSyncJob(URI artifactID, URI resourceID, StorageAdapter storageAdapter, ArtifactDAO artifactDAO) {
+        InventoryUtil.assertNotNull(FileSyncJob.class, "artifactID", artifactID);
+        InventoryUtil.assertNotNull(FileSyncJob.class, "resourceID", resourceID);
+        InventoryUtil.assertNotNull(FileSyncJob.class, "storageAdapter", storageAdapter);
+        InventoryUtil.assertNotNull(FileSyncJob.class, "artifactDAO", artifactDAO);
+
+        this.artifactID = artifactID;
+        this.resourceID = resourceID;
+        this.storageAdapter = storageAdapter;
+        this.artifactDAO = artifactDAO;
+    }
+
+    @Override
+    public void run() {
+        RegistryClient regClient = new RegistryClient();
+        log.debug("resource id: " + this.resourceID);
+        URL certURL = regClient.getServiceURL(this.resourceID, Standards.VOSPACE_SYNC_21, AuthMethod.CERT);
+        log.debug("certURL: " + certURL);
+
+        try {
+            // Ask for all protocols available back, and the server will
+            // give you URLs for the ones it allows.
+            List<Protocol> protocolList = new ArrayList<Protocol>();
+            Protocol httpsCert = new Protocol(VOS.PROTOCOL_HTTPS_GET);
+
+            log.debug("protocols: " + httpsCert);
+            protocolList.add(httpsCert);
+
+            Transfer transfer = new Transfer(artifactID, Direction.pullFromVoSpace, protocolList);
+            transfer.version = VOS.VOSPACE_21;
+
+            TransferWriter writer = new TransferWriter();
+            ByteArrayOutputStream out = new ByteArrayOutputStream();
+            writer.write(transfer, out);
+            FileContent content = new FileContent(out.toByteArray(), "text/xml");
+            log.debug("xml file content to be posted: " + transfer);
+
+            log.debug("artifact path: " + artifactID.getPath());
+            HttpPost post = new HttpPost(certURL, content, true);
+            post.prepare();
+            log.debug("post prepare done");
+
+            // TODO: error handling is likely very different than this...
+            if (post.getThrowable() != null && post.getThrowable() instanceof FileNotFoundException) {
+                throw (FileNotFoundException) post.getThrowable();
+            }
+
+            TransferReader reader = new TransferReader();
+            Transfer t = reader.read(post.getInputStream(), null);
+            List<String> urlList = t.getAllEndpoints();
+
+            log.debug("endpoints returned: " + urlList);
+            // get the URL list from t, then call HttpGet to get the item
+
+            int retryCount = 0;
+            boolean found = false;
+
+            for (int i = 0; i < urlList.size(); i++) {
+                URL u = new URL(urlList.get(i));
+                log.debug("trying this url: " + u);
+
```

Comment:
If this `while (retryCount` loop was outside the `for (URL u : urlList)` loop, it could be a good strategy for cases where some sites/servers are loaded to the point of failure and others are not... so not discounting it if those loops were swapped. As is, it tries a URL 2x with no delay and then moves on.

However, HttpGet already has a retry mechanism built in (it happens in `get.prepare()`) with a gradually increasing delay (unless the server says how long to delay) -- and delay between retries is absolutely necessary because otherwise the retries just make the situation worse instead of better. By default it only retries for transient exceptions, which is fine for now (it can be changed). The number of retries can be set (check code for default but I think it is in the ~3). 

The second thing about a retry loop is you have to identify when a retry might give a different result and when it will not. Here, I think that means a try/catch around the sequence from line 169 to 176 where you specifically deal with each exception that HttpGet.prepare() and StorageAdapter.put() might throw: 
* if the failure from prepare() is fatal (eg. ResourceNotFoundException ) then remove that URL from  the list
* if the failure from put() is fatal (eg WriteException or StorageEngageException?) then abort the whole job
Feel free to use multicatch style to handle several different exceptions in the same way.

This is the hard part because it is very hard to simulate (make a test that provokes) most of the failures so it comes down to thorough analysis (code review) to have good error handling.

Reply:
This led to a larger rework. The 'sync' part itself (HttpGet & storage adapter put) are in a function with a loop over the URL list. The outer while loop calls this function, and stops if the storage metadata is returned successfully, or its max number of retries is reached (in retrospect, maybe that number should be tuneable in the properties file?)

Split out the exceptions by retry or fail, and also by get or put to some degree. Maybe the latter isn't necessary as the log messages will have details. (?)